### ISSUE TITLE
Troubles printing data

### ISSUE THREAD

**Issue by alassaneg** [Raspberryy/Emulated_USB_Printer#2]
Hello,

First of all, thank you for your projects, it helps me a lot moving forward on my project.
Now, i'm having some issus printing data from my computer to the raspberry.
So, i've well followed the tutorial and my computer recognize the rasperry as a printer (in my case, i'm emulating a HP laserjet P1005).
When i launch the printing, the file goes to the printing queue and then nothing happens....I just receive the following characters : 12345x @PJL ENTER LANGUAGE=ACL (seems to be an hp command).
While testing this, did you have to send some kind of acknowledgment to the computer to start the printing?

Thank you in advance for your help.

**Comment by Raspberryy** [comment]
Hi,

If I remember correctly, I had similar issues with a different printer (Xerox maybe), when it was recognized correctly but wouldn't sent the print job. I never tried solving it. Also, I needed to send back data to PC. The original example of the gadget api contains a method for sending back data to the host ([here](https://www.kernel.org/doc/Documentation/usb/gadget_printer.rst)).

If you really need to stick to the `HP P1005` and keep the current driver you might want to use the USB traffic analyzer again: Start capturing traffic, connect a real `HP P1005`, send a very simple print job (empty page or one char) and check if your printer sends data on the BULK IN endpoint. If so, you can try copying that communication by sending the same data as the printer did. BUT this potentially requires an exact emulation of the original interface. The Raspberry Pi emulates an Endpoint 1 Bulk IN and Endpoint 1 Bulk OUT interface. If the original interface uses 1 Bulk OUT and 2 Bulk IN you might need to build your own version of the g_printer module (wouldn't want to that!). 

If you have access to the computer the pi is connected to, you can also try changing the drivers manually. That gives you the ability to for example select the `Generic Text-Only` driver on Windows which makes it much easier to work with the received data.

If you don't need to emulate that exact printer, try different ones! The foomatic DB contains almost 4000 printers. The most models I got working were **Ricoh Postscript printers.** Check [the Makefile](https://github.com/Raspberryy/Emulated_USB_Printer/blob/main/Makefile) for good examples.

It would be really cool if you can share the parameters for your HP LaserJet P1005 and any other printer you get working :)

Attached is a file which I got from a `HP1220` emulation. maybe this helps.
[hp1220_example.txt](https://github.com/Raspberryy/Emulated_USB_Printer/files/5928139/hp1220_example.txt)

**Comment by alassaneg** [comment]
Actually i don't have to stick with the HP 1005, i just took it randomly, but at the end i will have to stick with a HP or Epson printer. I just haven't found the parameters for the printers i want to emulate.

For now, i've tried  to emulate 3 different model and i can't get them working so i'll try to emulate a ricoh one as you suggest.
And i need to get myself a HP printer so i can try to analyze the USB traffic, if i can't find the information online. And hopefully, i won't have to develop my own driver!
But i actually don't think that the printer needs to send data to the computer. For example, HP uses the PJL (printer job language) to allow ones developping it's own application to communicate with the printer and return datas are not mentionned in the manual so the problem seems to be somewhere else, but i'll confirm that once i have a printer.

And actually, the HP LaserJet P1005 parameters are already in the PNP file you provided, i took it from there.

Thank you for your answer, it really helps!

**Comment by Raspberryy** [comment]
How are you going to process the received print jobs? I found, that GhostPDL isn't an optimal solution for converting PCL, which HP printers are most likely to have. That is another reason why used Ricoh printers. The best HP printer I found was the HP 1220. GhostPDL was able to convert it to readable PDF, but not able to extract the text.

Also be sure you have the correct VendorID and ProductID. I am not 100% sure but I think you need to set the HEX parameters with uppercase letters. so for HP use `idVendor=0x03F0` and not` idVendor=0x03f0`.

Again, if you have access to the Host, e.g. the computer the pi is connected to, you can simply change the drivers manually. That way you don't need to stick with PCL, ACL, EPS or ESC. I would suggest you use a PostScript printer if you want to extract data from the print jobs!

**Comment by alassaneg** [comment]
Actually, right now, i just want to check if i'm able to receive a print job whether it is PCL, ACL or anything else...So i'm not going to process anything right now. 
But at the end, i'll most likely have to stick with ESC/P printers and i'm planning on processing the received print jobs with the following tool : https://github.com/receipt-print-hq/escpos-tools . I might adapt it to your code.

Got it regarding the uppercase letters, i'm already using them.

I didn't know that changing the driver was an option so thank you for the info. I thought that one printer could work with a unique driver : the vendor driver.

**Comment by Raspberryy** [comment]
If you connect the real printer you will of course need the actual driver. So, if you are planning to forward the print job to the real printer someday you might want to stick to the real driver.

Are you now able to emulate a printer and receive some data? I didn't quite catch that.

**Comment by alassaneg** [comment]
Right now, i'm just checking if the g_printer is working.
I've tried to emulate the Ricoh Aficio SP35010SF but still not able to receive the print job...
Could you tell me which printer you've used that work?

**Comment by Raspberryy** [comment]
All listed in the Makefile (except the Kyocera FS1020D) were detected by my Windows 10 PC. The Ricoh printers worked best. 
Do you get any error message? Do you still only receive a few lines? 

I mentioned this in a different Issue but maybe this is your problem as well: The printer.c contains functions for setting three parameters: Paper Loaded/Paper Out, Error/ No Error, and Selected / Not selected. The correct status is Selected, Paper Loaded, No Error. Confirm that this is the case for you by running emulatedprinter -get_status. Change values if needed (See emulatedprinter -h).

Whats your host? What are you using to print? What drivers are currently loaded? What do you receive?

**Comment by alassaneg** [comment]
I don't get any error messages. 
Actually, i tried again and it seems to work well. I first tried to print this page but its a complicated one so i opened word and tried to print a single white page and it worked well, finally!

Now that the driver and the code are working, i need to get my self an ESC/P printer and it's pnpString (hard to find online so I might buy one).

I still need to figure out what is going on with these other printers. HP and epson don't mention any return values from the printer to the host in their commands reference manual...
I really appreciate your help.

**Comment by Raspberryy** [comment]
If you share I little mor of your setup, maybe we can help :) If you ever publish your project, keep me updated, I am really interested in seeing how people use this repo!

**Comment by alassaneg** [comment]
I finally got myself an epson TM-T88V printer.
For some reason, it doesn't have a pnp number. I didn't see it with usblyzer but i was able to get the vendor and product ID and after installing manually the driver, the printer work.
Then, when trying to emulate it, it's not working. Windows can see the emulated printer but it's considered as unavailable.
Actually the TM-T88V doesn't seem to act as a normal printer. When i plug it, the printer doesn't appear in the printer tree of the peripheral manager...
So the conclusion is that i don't think the gadget_printer driver can handle that POS printer.

**Comment by Raspberryy** [comment]
Thats weird :/ You can share the USBlyzer capture file, maybe I can help you :)

**Comment by alassaneg** [comment]
[TM-88V_LOG.zip](https://github.com/Raspberryy/Emulated_USB_Printer/files/5972398/TM-88V_LOG.zip)

The file is right here!

**Comment by Raspberryy** [comment]
Looks like the best solution is getting started on building your own version of the g_printer module. It looks quite similar to the emulated interface of the g_printer module (e.g. same Endpoints), but the interface descriptor is different. You might simply need to change bInterfaceClass and the bInterfaceSubClass for your version of the g_printer module. Sadly I never needed to build a custom version and currently I don't really have the time to get into this. 

But hey, that might be a cool project!

**Comment by alassaneg** [comment]
That was my concern but anyway, i'll still have to give a try then!
I'll have to go deep in the g_printer and i'll try to modify it.

Thanks again for your help

**Comment by alassaneg** [comment]
Hello,

I'm back for one more question for you!
First of all, i was using the TM88 printer using the vendor device class and i was able to change it using the printer class by changing one option in the printer utility software.
I then figured out that the driver was sending vendor specific request to the printer so i had to modify and rebuild the g_printer driver so he knows about these vendor request but unfortunately, i was still unable to catch the data using g_printer.
So, i switched to the generic/text only driver and it works perfectly when using the TM88 printer but once again, doesn't work with g_printer...
Have you ever tried to use g_printer using the generic text only driver?

**Comment by Raspberryy** [comment]
Yes, I did. Generic Text Only were my first attempts and worked perfectly fine. 
Do I understand correctly, that you compiled your own kernel module, aka "rebuild the g_printer"? Or did you adjusted the code talking to /etc/g_printer?
Also, your last sentence is at least for me a little bit confusing. Do you refer to the emulated model?

**Comment by alassaneg** [comment]
No, i've just added a few lines to the driver source code ( https://elixir.bootlin.com/linux/v5.4.41/source/drivers/usb/gadget/function/f_printer.c ) and recompiled it.
Regarding the last sentence, indeed, i emulated the TM88V printer and tried to use it with the Generic Text Only and it was not working...
What is the model of the printer you've emulated to make it work with the Generic Text Only ?

**Comment by Raspberryy** [comment]
I was able to use multiple printers with the Generic Text Only. All I did was

1. Find parameters for g_printer module by looking at the enumeration of a real printer/looking in the FoomaticDB
2. Connecting the Pi to a PC
3. Starting g_printer with needed parameters
4. Confirm PC detects Pi as the printer model that needs to be emulated
5. Go to Printer Settings on PC
6. Manually change driver to Generic Text Only
7. Run printer.c from this repository
8. Print on PC

That way the PC uses the Text Only driver. For printing I only used MS Word and notepad.exe.
Maybe the programm you are using to print does not accept the drivers..?

**Comment by alassaneg** [comment]
Ok now it's working.
So what i did first was:
1- Pluging the TM88 printer
2- Add a new printer
3- use the generic/text printer
4- print
That worked perfect.

After that, i unplugged the printer and i :
5- loaded g_printer
6- Plugged the embedded linux (not  raspberry in my case but an atsama5d27 sm1 ek)
7-  Launche printer.c and tried to print to the embedded linux but it  doesn't seem to be fully recognize as a tm88 printer so i repeated steps 2 to 4 and added once again a new printer and then it worked.

So, it's not really what i want because for my final product, i'm going to use a usb hub to redirect the printjob to the embedded linux or to the printer so i need windows to believe that my embedded linux board is a real tm88 printer so i need to figure out what is going on!

**Comment by Raspberryy** [comment]
Awesome. My guess would still be that something in the descriptors/Device ID is wrong. Did you see that the g_printer module trims the device ID? E.g. when ipnpstring="12345", the g_printer module will send "123", so you might need to call the module with "12345  " ("12345" + 2 spaces).

**Comment by alassaneg** [comment]
I just took a look at the usblyzer logs and indeed, the two last characters of the pnp string are missing, i'd have never noticed that so thanks a lot for the info! How did you notice that?
I'll make more test tomorrow when i'll have access to the printer.
But when using the ricoh aficio pnp string without leaving a sapce at the end it was working, but i'll make sure that the issue is still coming from that pnp string.

**Comment by Raspberryy** [comment]
I can't remember when or how I noticed that. My first thought was, maybe USBlyzer is wrong, because all printer models worked just fine without the double spaces. Maybe Windows only parses MFG and MDL, so it doesn't matter if CMD is formatted correctly or not. But this is just a guess. 
I would be curious if this would solve one of your problems.

**Comment by alassaneg** [comment]
It'ill probably solve one of my issue, i'll let you know as soon as i can test this.
I also noticed something "interesting". 
The tm88 printer driver sends periodically an "in 66048 bytes bulk transfer" to the gadget driver. Everytime, it fails.
The ricoh printer driver sends periodically an "in 512 bytes bulk transfer" to the gadget driver and it's a success everytime. 
So right now, i'm looking at the bulk capabilities of the gadget driver.

**Comment by alassaneg** [comment]
New updates!
I figured out that i needed to set the iManufacturer and iProduct so windows could then recognize the embedded linux as the real printer and load th driver as if it was the TM88V that was plugged in the computer.
However, the bulk in request stilll fail and i'm still investigating on that.

**Comment by Raspberryy** [comment]
So you are increasing buffers to recreate the message?

**Comment by alassaneg** [comment]
So, i did it, i increased the buffer size but it's still not working.
I still need to resolve this, but for now, i installed the gadget printer manually on windows and in the printer parameters, i've selected the espon APD5 driver and i'm able to receive the printer datas on linux.
It's not exaclty what i want, but it's gonna be ok for now. But i'm still not done with this bulk transfer issue, i really need to understant what is going on.

**Comment by RWAP** [comment]
> 
> 
> So, i did it, i increased the buffer size but it's still not working.
> I still need to resolve this, but for now, i installed the gadget printer manually on windows and in the printer parameters, i've selected the espon APD5 driver and i'm able to receive the printer datas on linux.
> It's not exaclty what i want, but it's gonna be ok for now. But i'm still not done with this bulk transfer issue, i really need to understant what is going on.

Did you get any further with the TM88V being recognised?

I have actually written a little bit of code which waits until an ESC character is received (which indicates the start of an ESC/POS sequence).  Anything received up until then is echoed to the USB printer, and the response is captured and sent back through the gadget printer to the host PC.

This should work in theory, but Windows still does not recognise the gadget printer as the TM88V.

**Comment by bcheidemann** [comment]
@RWAP any chance you could share the code? Would be super interested to see it!

**Comment by alassaneg** [comment]
> > So, i did it, i increased the buffer size but it's still not working.
> > I still need to resolve this, but for now, i installed the gadget printer manually on windows and in the printer parameters, i've selected the espon APD5 driver and i'm able to receive the printer datas on linux.
> > It's not exaclty what i want, but it's gonna be ok for now. But i'm still not done with this bulk transfer issue, i really need to understant what is going on.
> 
> Did you get any further with the TM88V being recognised?
> 
> I have actually written a little bit of code which waits until an ESC character is received (which indicates the start of an ESC/POS sequence). Anything received up until then is echoed to the USB printer, and the response is captured and sent back through the gadget printer to the host PC.
> 
> This should work in theory, but Windows still does not recognise the gadget printer as the TM88V.

No, i'm still not able to get windows recognize the gadget printer as the TM88V. As i said, my workaround was to manually install the gadget printer and then to select the APD5 driver in the printer options.

**Comment by Raspberryy** [comment]
Hi, I am curious, why are all of you interested in the TM88V? Is there anything special about that printer?

**Comment by RWAP** [comment]
> 
> 
> Hi, I am curious, why are all of you interested in the TM88V? Is there anything special about that printer?

Nothin in particular - it just seems to be a widely used printer in POS (Point of Sale) applications

**Comment by RWAP** [comment]
Code for reading printer query - sending it to the printer and responding (may be some bugs here as I have had to chop it out of my larger project):

```
  static int
  read_printer_data()
  {
	// Print Info
	if(display_printer_status() != 0)
		return -1;
	printf("[*] Start reading printjobs\n");

	/* Open device file for printer gadget. */
	struct pollfd fd[1], fusb[1]; // RWAP
	fd[0].fd = open(PRINTER_FILE, O_RDWR);
	if (fd[0].fd < 0) {
		printf("[!] Error\t%d opening %s\n", fd[0].fd, PRINTER_FILE);
		close(fd[0].fd);
		return(-1);
	}

	fd[0].events = POLLIN | POLLRDNORM;
    int esc = 0, device_query_bytes = 0, printer_response_bytes = 0, response_echo = 0, usb_printer_port_open = 0; // RWAP
    int response_timeover = 0, response_received = 0;  // RWAP
	int rec = -1;
	char filename[100];
	char pjob_outp_file[512];
	FILE* cur_job;

	while (1) {
		static char buf[BUF_SIZE];
		int bytes_read;
		int retval;

		/* Wait for up to 1 second for data. */
		retval = poll(fd, 1, 1000);

		if (retval && (fd[0].revents & POLLRDNORM)) {

			/* Data received - Create new file if needed */
			if(rec == -1) {
                        	time_t     now;
                        	struct tm  ts;

                        	// Build file name
                        	time(&now);
                        	ts = *localtime(&now);
                        	strftime(filename, sizeof(filename), "%Y-%m-%d-%H-%M-%S", &ts);
                        	sprintf(pjob_outp_file,"%s%s.%s", FILE_OUTPUT_PATH, filename, PRINTJOB_FILE_EXTENSION);

                        	// Create PCL file
                        	cur_job = fopen(pjob_outp_file, "w+");
                        	if(cur_job == NULL){
                                	printf("[!] Error\tOpening %s. Exiting\n", pjob_outp_file);
                                	return -1;
                        	}

                        	rec = 0;
			}


			/* Read data from printer gadget driver. */
            device_query_bytes = 0; // RWAP
            // lseek(fd[0].fd,0,SEEK_SET); // Wind back to start then try to read             // RWAP
			bytes_read = read(fd[0].fd, buf, BUF_SIZE);

			if (bytes_read < 0) {
				printf("[!] Error\t%d reading from %s\n",
						fd[0].fd, PRINTER_FILE);
				close(fd[0].fd);
				return(-1);                
            } else if (bytes_read > 0) {
                // Changed by RWAP
                if (printer_response_bytes && bytes_read == printer_response_bytes) {                        
                    response_echo = 1;
                    // Check if this is the response we just sent
                    for (i=0; i<printer_response_bytes; i++) {
                        printf("Response Check = %d = %d \n",buf[i],printer_response[i]);                            
                        if (buf[i] != printer_response[i]) {                                
                            response_echo = 0;
                        }
                    }                                                
                } else {
                    response_echo = 0;
                }
                if (!response_echo) {
                    printer_response_bytes = 0;
                    if (esc==0) {
                        t=0;
                        device_query_bytes = bytes_read;
                        for (i=0; i<bytes_read; i++) {
                            // Strip back any esc characters - as that signifies the start of ESC/POS printing
                            if (buf[i] == 27) {
                                if (i > 0) {
                                    device_query_bytes = i - 1;
                                    break;
                                } else {
                                    device_query_bytes = 0;
                                    break;
                                }
                            }
                        }
                        if (device_query_bytes) {
                            if (!usb_printer_port_open) {                            
                                /* Open device file for printer gadget. */
                                fusb[0].fd = open("/dev/usb/lp0", O_RDWR|O_NONBLOCK);
                                if (fusb[0].fd < 0) {
                                    printf("Error %d opening %s\n", fusb[0].fd, RASPBERRYPI_USB_PRINTER);
                                    close(fusb[0].fd);
                                    return(-1);
                                }            
                                fusb[0].events = POLLIN | POLLRDNORM;
                                usb_printer_port_open = 1; 
                            }                                                          
                            
                            // Send device query to printer
                            for (i=0; i<device_query_bytes; i++) {
                                printer_query[i] = buf[i];
                            }
                            // Write Query to USB Printer
                            write (fusb[0].fd, printer_query, device_query_bytes);                                                        
                            
                            // Fetch result from printer
                            /* Allows 30 seconds for response. */
                            response_timeover = 0;
                            response_received = 0;                                
                            while (!response_received && response_timeover < 30 * 4) {                                
                                /* Wait for up to 1/4 second for data. */
                                retval = poll(fusb, 1, 250);
                                response_timeover++;
                                if (retval && (fusb[0].revents & (POLLIN | POLLRDNORM))) {
                                    /* Read data from printer gadget driver. */
                                    printer_response_bytes = read(fusb[0].fd, usb_buf, BUF_SIZE);
                                    if (printer_response_bytes) {
                                        // Send device query response to OTG port
                                        for (i=0; i<printer_response_bytes; i++) {
                                            printer_response[i] = usb_buf[i];                                                
                                        }
                                        write(fd[0].fd, printer_response, printer_response_bytes);
                                        response_received = printer_response_bytes;
                                    }
                                }
                            }
                        } else {
                            if (usb_printer_port_open) {
                                // Close the device query to the USB printer
                                close(fusb[0].fd);
                                usb_printer_port_open = 0;
                            }
                        }
                        if (bytes_read > device_query_bytes) {                            
                            t=0;
                            for (i=0; i<bytes_read; i++) {
                                // We ignore anything at the start of incoming data until an ESC is found
                                // Valid for ESC/P2, ESC/POS and PCL code
                                if (buf[i]==27) {
                                    esc=1;
                                    printer_response_bytes = 0;
                                }
                                if (esc==1) {
                                    buf2[t]=buf[i];
                                    t++;
                                }
                            }
                            if (esc==1) {
                                rec = 1;
                                /* Write data to standard OUTPUT (stdout). */
                                fwrite(buf, 1, bytes_read, cur_job);
                                fflush(stdout);
                            }
                        }                            
                    }
                } else {
                    rec = 1;
                    /* Write data to standard OUTPUT (stdout). */
                    fwrite(buf, 1, bytes_read, cur_job);
                    fflush(stdout);
                }
            } else {
                // sleep_ms(100);
            }                


		} else if(rec == 1) {
			rec = -1;
			fclose(cur_job);

			// Start async printjob conversion
			if(fork() == 0){
				gpdl_parse_printjob(pjob_outp_file, filename);
			}

			printf("[*] FINSIEHD\tReceived print job %s.%s\n",
                        	filename, PRINTJOB_FILE_EXTENSION);
		}
	}

	/* Close the device file. */
	close(fd[0].fd);

	return 0;
  }
```

**Comment by Raspberryy** [comment]
Hi All :) For anyone still looking into this, I found this by accident and thought it might be worth sharing it here. Not to sure if this repo is actually working or not, but anyway here a link to [a fork related to the TM88 printer](https://github.com/andrey5608/raspberry-pi-printer-emulator/tree/epson-tm88-emulator)
